// @framer/jakobjordan.spring-numbers/code/SpringNumbers.tsx
import * as React from "react"
import { Frame, addPropertyControls, ControlType } from "framer"
import { useSpring, animated, config } from "react-spring"
import { Spring } from "react-spring/renderprops"

// Open Preview: Command + P
// Learn more: https://framer.com/api

export function SpringNumber(props) {
    const {
        tint,
        suffix,
        fontSize,
        fontFamily,
        decimal,
        initialNumber,
        duration,
        ...rest
    } = props

    const style = {
        fontFamily,
        fontSize,
        background: "none",
        color: tint,
        width: 120,
        height: 70,
    }

    const numbers = useSpring({
        number: props.initialNumber,
        from: { number: props.initialNumber },
        config: { duration: props.duration },
    })

    return (
        <Frame style={style}>
            <animated.span>
                {numbers.number.interpolate(val => val.toFixed(props.decimal))}
            </animated.span>
            {suffix}
        </Frame>
    )
}

SpringNumber.defaultProps = {
    height: 128,
    width: 240,

    tint: "#0099ff",
    suffix: "°",
    initalNumber: 60,
    fontSize: 50,
    fontFamily: "Arial",
}

// Learn more: https://framer.com/api/property-controls/
addPropertyControls(SpringNumber, {
    initialNumber: {
        title: "Number",
        type: ControlType.Number,
        defaultValue: 60,
        step: 0.1,
    },
    decimal: {
        title: "Decimal",
        type: ControlType.Number,
        defaultValue: 1,
    },
    duration: {
        title: "Duration",
        type: ControlType.Number,
        defaultValue: 500,
        min: 0,
        max: 60000,
        step: 50,
    },
    fontSize: {
        title: "Font Size",
        type: ControlType.Number,
        defaultValue: 60,
    },

    suffix: {
        title: "Suffix",
        type: ControlType.String,
        defaultValue: "",
    },
    fontFamily: {
        title: "Font",
        type: ControlType.String,
        defaultValue: "Arial",
    },
    tint: {
        title: "Tint",
        type: ControlType.Color,
        defaultValue: "#444",
    },
})
